src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -91,6 +106,94 @@ public EETypeNode(NodeFactory factory, TypeDesc type)
+                // the current type interesting for GVM analysis (i.e. instantiate its overriding GVMs for existing GVMDependenciesNodes
+                // of the instantiated GVM on the parent types).
+                if (method.HasInstantiation)
+                {
+                    result |= VirtualMethodAnalysisFlags.NeedsGvmEntries;
+
+                    MethodDesc slotDecl = MetadataVirtualMethodAlgorithm.FindSlotDefiningMethodForVirtualMethod(method);
+                    if (slotDecl != method)
+                        result |= VirtualMethodAnalysisFlags.InterestingForDynamicDependencies;
+                }
+
+                // Early out if we set all the flags we could have set
+                if ((result & VirtualMethodAnalysisFlags.AllFlags) == VirtualMethodAnalysisFlags.AllFlags)
+                    return result;
+            }
+
+            //
+            // Check if the type implements any interface, where the method implementations could be on
+            // base types.
+            // Example:
+            //      interface IFace {
+            //          void IFaceGVMethod<U>();
+            //      }
+            //      class BaseClass {
+            //          public virtual void IFaceGVMethod<U>() { ... }
+            //      }
+            //      public class DerivedClass : BaseClass, IFace { }
+            //
+            foreach (DefType interfaceImpl in defType.RuntimeInterfaces)
+            {
+                foreach (MethodDesc method in interfaceImpl.GetAllVirtualMethods())
+                {
+                    if (!method.HasInstantiation)
+                        continue;
+
+                    // We found a GVM on one of the implemented interfaces. Find if the type implements this method.
+                    // (Note, do this comparison against the generic definition of the method, not the specific method instantiation
+                    MethodDesc slotDecl = method.Signature.IsStatic ?
+                        defType.ResolveInterfaceMethodToStaticVirtualMethodOnType(method)
+                        : defType.ResolveInterfaceMethodTarget(method);
+                    if (slotDecl != null)
+                    {
+                        // If the type doesn't introduce this interface method implementation (i.e. the same implementation
+                        // already exists in the base type), do not consider this type interesting for GVM analysis just yet.
+                        //
+                        // We need to limit the number of types that are interesting for GVM analysis at all costs since
+                        // these all will be looked at for every unique generic virtual method call in the program.
+                        // Having a long list of interesting types affects the compilation throughput heavily.
+                        if (slotDecl.OwningType == defType ||
+                            defType.BaseType.ResolveInterfaceMethodTarget(method) != slotDecl)
```

Comment:
What's an example of when this could happen?

My understanding of the code is that it's checking to see if the impl of the interface in the base type is different, but only when the current impl isn't declared in the current type. That would seem to imply to me that the impl is somehow being pulled from above the base type, which seems impossible at first glance.

Reply:
Example is a twist on the sample 30 lines above.

```
            //      interface IFace {
            //          void IFaceGVMethod<U>();
            //      }
            //      class BaseClass {
            //          public virtual void IFaceGVMethod<U>() { ... }
            //      }
            //      public class DerivedClass : BaseClass, IFace { }
            //
```

Move the `IFace` implementation down to `BaseClass`. Now if we try to resolve `IFaceGVMethod` on `DerivedClass` we get a method from the base, but it already implemented the interface in the base. So we don't have to think about it in the context of `DerivedClass`. Removing DerivedClass from the N*M algorithm is a compile throughput improvement.

(This is a copy from one of the places that I'm unifying - not new code).

Reply:
Sorry, still not getting it. If the example code is 

```C#
            //      interface IFace {
            //          void IFaceGVMethod<U>();
            //      }
            //      class BaseClass : IFace {
            //          public virtual void IFaceGVMethod<U>() { ... }
            //      }
            //      public class DerivedClass : BaseClass { }
            //
```

I would have assumed that `slotDecl.OwningType` would be `IFace`, and `defType` would be `BaseClass`. Is `slotDecl` the implementing slot?

Reply:
`defType` would be `DerivedClass`.

We go over all interfaces on `DerivedClass`. For each interface method, we resolve to the implementation. If the implementation is something else than it was on the base, we need to track it. If the implementation is the same thing that it was on the base, we don't need to track it.

The `if (slotDecl.OwningType == defType)` part of the check is just a perf optimization to avoid going to the resolve step on base. If the implementation is on the type we're looking at right now, for sure the base type wouldn't have that.